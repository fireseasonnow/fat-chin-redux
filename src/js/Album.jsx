import React from 'react';
import styled from 'styled-components';
import Track from './Track';

const Album = ({
    name,
    coverUrl,
    year,
    tracks,
    tracksVisible,
    index,
    tracksVisibilityHandler,
    coverZoomHandler,
    lyricsVisibilityHandler,
}) => (
    <Item>
        <Title>
            {`${name} (${year})`}
        </Title>
        <CoverWrapper>
            <CoverImg src={coverUrl} alt={`'${name}' album cover`} />
            <Overlay>
                <OverlayItem onClick={() => tracksVisibilityHandler(index)} type='button'>{tracksVisible ? 'Hide tracks' : 'Show tracks'}</OverlayItem>
                <OverlayItem onClick={() => coverZoomHandler(index)} type='button'>Zoom cover</OverlayItem>
            </Overlay>
        </CoverWrapper>

        <Tracks>
            {tracksVisible && tracks.map((track, trackIndex) => (
                <Track
                    key={trackIndex}
                    number={parseInt(trackIndex, 10) + 1}
                    name={track}
                    lyricsVisibilityHandler={lyricsVisibilityHandler}
                />
            ))}
        </Tracks>
    </Item>
);

export default Album;

const Item = styled.li`
    list-style: none;
`;

const Title = styled.h2`
    margin: 0;
    font-size: 1em;
    text-align: center;
`;

const CoverImg = styled.img`
    display: block;
    max-width: 100%;
    height: auto;
    filter: grayscale(1);
    transition: filter ease-in-out 250ms;

    ${Item}:hover & {
        filter: grayscale(0);
        transition: filter ease-in-out 250ms;
    }
`;

const CoverWrapper = styled.div`
    position: relative;
`;

const Tracks = styled.ul`
    padding-left: 0;
`;

const Overlay = styled.div`
    display: grid;
    grid-template-columns: 1fr 1fr;
    position: absolute;
    bottom: 0;
    width: 100%;
    height: 50px;
    background: rgba(255, 255, 255, 0.7);
    opacity: 0;
    transition: opacity ease-in-out 250ms;

    ${CoverWrapper}:hover & {
        opacity: 1;
        transition: opacity ease-in-out 250ms;
    }
`;

const OverlayItem = styled.button`
    display: flex;
    justify-content: center;
    align-items: center;
    text-decoration: none;
    color: #000;
    background: rgba(255, 255, 255, 0.05);
    cursor: pointer;
    font-family: 'Oswald', sans-serif;
    font-size: 1em;
    border: 0;

    &:first-child,
    &:first-child:active {
        border-right: 1px solid #999;
    }
`;
