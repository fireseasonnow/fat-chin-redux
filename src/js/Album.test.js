import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallowToJson } from 'enzyme-to-json';

import Album from './Album';

configure({ adapter: new Adapter() });

describe('<Album />', () => {
    let wrapper = {};

    beforeEach(() => {
        wrapper = shallow(<Album />);
    });

    it('should render without crashing', () => {
        expect(wrapper).toBeTruthy();
    });

    it('should render an album without tracks and \'Show tracks\' text', () => {
        wrapper.setProps({
            name: 'Hello',
            coverUrl: 'url',
            year: 1984,
            tracksVisible: false,
        });

        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('should render an album with tracks and \'Hide tracks\' text', () => {
        wrapper.setProps({
            name: 'Hello',
            coverUrl: 'url',
            year: 1984,
            tracksVisible: true,
            tracks: [
                'blah',
                'blah-blah'
            ],
        });

        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('should fire tracksVisiblityHandler when clicked on a tracks button', () => {
        const tracksVisibilityHandler = jest.fn();

        wrapper.setProps({
            index: 4,
            tracksVisibilityHandler,
        });

        wrapper.find('Album__OverlayItem:first-child').simulate('click');

        expect(tracksVisibilityHandler).toHaveBeenCalledWith(4);
    });

    it('should fire coverZoomHandler when clicked on a zoom button', () => {
        const coverZoomHandler = jest.fn();

        wrapper.setProps({
            index: 3,
            coverZoomHandler,
        });

        wrapper.find('Album__OverlayItem:last-child').simulate('click');

        expect(coverZoomHandler).toHaveBeenCalledWith(3);
    });
});
