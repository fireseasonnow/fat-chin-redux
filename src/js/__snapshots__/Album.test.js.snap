// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Album /> should render an album with tracks and 'Hide tracks' text 1`] = `
<Album__Item>
  <Album__Title>
    Hello (1984)
  </Album__Title>
  <Album__CoverWrapper>
    <Album__CoverImg
      alt="'Hello' album cover"
      src="url"
    />
    <Album__Overlay>
      <Album__OverlayItem
        onClick={[Function]}
        type="button"
      >
        Hide tracks
      </Album__OverlayItem>
      <Album__OverlayItem
        onClick={[Function]}
        type="button"
      >
        Zoom cover
      </Album__OverlayItem>
    </Album__Overlay>
  </Album__CoverWrapper>
  <Album__Tracks>
    <Track
      key="0"
      name="blah"
      number={1}
    />
    <Track
      key="1"
      name="blah-blah"
      number={2}
    />
  </Album__Tracks>
</Album__Item>
`;

exports[`<Album /> should render an album without tracks and 'Show tracks' text 1`] = `
<Album__Item>
  <Album__Title>
    Hello (1984)
  </Album__Title>
  <Album__CoverWrapper>
    <Album__CoverImg
      alt="'Hello' album cover"
      src="url"
    />
    <Album__Overlay>
      <Album__OverlayItem
        onClick={[Function]}
        type="button"
      >
        Show tracks
      </Album__OverlayItem>
      <Album__OverlayItem
        onClick={[Function]}
        type="button"
      >
        Zoom cover
      </Album__OverlayItem>
    </Album__Overlay>
  </Album__CoverWrapper>
  <Album__Tracks />
</Album__Item>
`;
